File: src/api/git.rs
use anyhow::Context;
use colored::*;
use git2::Repository;
use crate::errors::{DevFlowError, Result};

pub struct GitClient {
    repo: Repository,
}

impl GitClient {
    pub fn new() -> Result<Self> {
        let repo = Repository::open_from_env()
            .map_err(|_| DevFlowError::NotInGitRepo)?;

        Ok(Self { repo })
    }

    pub fn is_clean(&self) -> Result<bool> {
        let statuses = self.repo.statuses(None)
            .map_err(|e| DevFlowError::Other(format!("Failed to get git status: {}", e)))?;
        Ok(statuses.is_empty())
    }

    pub fn current_branch(&self) -> Result<String> {
        let head = self.repo.head()
            .map_err(|e| DevFlowError::Other(format!("Failed to get HEAD reference: {}", e)))?;

        if head.is_branch() {
            let branch_name = head
                .shorthand()
                .ok_or_else(|| DevFlowError::Other("Branch name contains invalid UTF-8".to_string()))?
                .to_string();

            Ok(branch_name)
        } else {
            Err(DevFlowError::Other("Not currently on a branch (detached HEAD state)".to_string()))
        }
    }

    pub fn create_branch(&self, branch_name: &str) -> Result<()> {
        let head_commit = self
            .repo
            .head()
            .context("Failed to get HEAD")?
            .peel_to_commit()
            .context("Failed to get HEAD commit")?;

        self.repo
            .branch(branch_name, &head_commit, false)
            .context(format!("Failed to create branch '{}'", branch_name))?;

        let refname = format!("refs/heads/{}", branch_name);

        self.repo
            .set_head(&refname)
            .context("Failed to set HEAD to new branch")?;

        self.repo
            .checkout_head(None)
            .context("Failed to checkout new branch")?;

        println!(
            "{}",
            format!("✓ Created and switched to branch '{}'", branch_name).green()
        );

        Ok(())
    }

    pub fn status_summary(&self) -> Result<String> {
        let statuses = self.repo.statuses(None)
            .map_err(|e| DevFlowError::Other(format!("Failed to get git status: {}", e)))?;

        if statuses.is_empty() {
            return Ok("  Working directory clean".to_string());
        }

        let mut summary = Vec::new();

        for entry in statuses.iter() {
            let status = entry.status();
            let path = entry.path().unwrap_or("unknown");

            if status.is_wt_modified() {
                summary.push(format!("  {} {}", "M".yellow(), path));
            } else if status.is_wt_new() {
                summary.push(format!("  {} {}", "A".green(), path));
            } else if status.is_wt_deleted() {
                summary.push(format!("  {} {}", "D".red(), path));
            }
        }

        Ok(summary.join("\n"))
    }

    pub fn push(&self, branch_name: &str) -> Result<()> {
        let mut remote = self
            .repo
            .find_remote("origin")
            .context("Failed to find remote 'origin'")?;

        let refspec = format!("refs/heads/{}:refs/heads/{}", branch_name, branch_name);

        let mut callbacks = git2::RemoteCallbacks::new();
        callbacks.credentials(|_url, username_from_url, _allowed_types| {
            git2::Cred::ssh_key_from_agent(username_from_url.unwrap_or("git"))
        });

        let mut push_options = git2::PushOptions::new();
        push_options.remote_callbacks(callbacks);

        remote
            .push(&[&refspec], Some(&mut push_options))
            .context(format!("Failed to push branch '{}'", branch_name))?;

        println!(
            "{}",
            format!("✓ Pushed branch '{}' to origin", branch_name).green()
        );

        Ok(())
    }

    pub fn commit(&self, message: &str) -> Result<()> {
        let mut index = self.repo.index().context("Failed to get repository index")?;
        index.add_all(["*"].iter(), git2::IndexAddOption::DEFAULT, None)
            .context("Failed to add files to index")?;
        index.write().context("Failed to write index")?;

        let tree_id = index.write_tree().context("Failed to write tree")?;
        let tree = self.repo.find_tree(tree_id).context("Failed to find tree")?;

        let head = self.repo.head().context("Failed to get HEAD")?;
        let parent_commit = head.peel_to_commit().context("Failed to get parent commit")?;

        let signature = self.repo.signature()
            .context("Failed to get git signature. Make sure git user.name and user.email are configured")?;

        self.repo
            .commit(
                Some("HEAD"),
                &signature,
                &signature,
                message,
                &tree,
                &[&parent_commit],
            )
            .context("Failed to create commit")?;

        println!("{}", format!("✓ Created commit: {}", message).green());

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_git_client_in_repo() {
        let result = GitClient::new();
        assert!(result.is_ok(), "Should successfully open git repo");
    }

    #[test]
    fn test_current_branch() {
        if let Ok(git) = GitClient::new() {
            let result = git.current_branch();
            match result {
                Ok(branch) => assert!(!branch.is_empty(), "Branch name should not be empty"),
                Err(_) => {
                    // It's okay if we can't get branch (e.g., new repo with no commits)
                }
            }
        }
    }

    #[test]
    fn test_status_summary() {
        if let Ok(git) = GitClient::new() {
            let result = git.status_summary();
            assert!(result.is_ok(), "Should get status summary");
        }
    }
}

File: internal/git/git.go
package git

import (
	"bytes"
	"fmt"
	"os/exec"
	"strings"

	"github.com/Ilia01/devflow/internal/utils"
)

type Client struct {
	worktree string
}

func NewClient() (*Client, error) {
	out, err := runInDir("", "rev-parse", "--show-toplevel")
	if err != nil {
		return nil, fmt.Errorf("not in git repository: %w", err)
	}
	return &Client{worktree: strings.TrimSpace(out)}, nil
}

func (c *Client) CurrentBranch() (string, error) {
	out, err := runInDir(c.worktree, "rev-parse", "--abbrev-ref", "HEAD")
	if err != nil {
		return "", err
	}
	branch := strings.TrimSpace(out)
	if branch == "HEAD" {
		return "", fmt.Errorf("detached HEAD state")
	}
	return branch, nil
}

func (c *Client) IsClean() (bool, error) {
	out, err := runInDir(c.worktree, "status", "--porcelain")
	if err != nil {
		return false, err
	}
	return strings.TrimSpace(out) == "", nil
}

func (c *Client) StatusSummary() (string, error) {
	out, err := runInDir(c.worktree, "status", "--short")
	if err != nil {
		return "", err
	}
	trimmed := strings.TrimSpace(out)
	if trimmed == "" {
		return "  Working directory clean", nil
	}
	var lines []string
	for _, line := range strings.Split(trimmed, "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		prefix := ""
		if len(line) >= 2 {
			status := strings.TrimSpace(line[:2])
			switch {
			case strings.Contains(status, "M"):
				prefix = utils.Yellow("M")
			case strings.Contains(status, "A"):
				prefix = utils.Green("A")
			case strings.Contains(status, "D"):
				prefix = utils.Red("D")
			default:
				prefix = status
			}
			path := strings.TrimSpace(line[2:])
			lines = append(lines, fmt.Sprintf("  %s %s", prefix, path))
		} else {
			lines = append(lines, "  "+line)
		}
	}
	return strings.Join(lines, "\n"), nil
}

func (c *Client) CreateBranch(branch string) error {
	_, err := runInDir(c.worktree, "checkout", "-b", branch)
	return err
}

func (c *Client) Push(branch string) error {
	_, err := runInDir(c.worktree, "push", "-u", "origin", branch)
	return err
}

func (c *Client) Commit(message string) error {
	if _, err := runInDir(c.worktree, "add", "-A"); err != nil {
		return err
	}
	cmd := exec.Command("git", "commit", "-m", message)
	cmd.Dir = c.worktree
	var stderr bytes.Buffer
	cmd.Stdout = &stderr
	cmd.Stderr = &stderr
	if err := cmd.Run(); err != nil {
		return fmt.Errorf("git commit failed: %s", strings.TrimSpace(stderr.String()))
	}
	return nil
}

func (c *Client) Root() string {
	return c.worktree
}

func runInDir(dir string, args ...string) (string, error) {
	cmd := exec.Command("git", args...)
	if dir != "" {
		cmd.Dir = dir
	}
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr
	if err := cmd.Run(); err != nil {
		if stderr.Len() > 0 {
			return "", fmt.Errorf("%s", strings.TrimSpace(stderr.String()))
		}
		return "", err
	}
	return stdout.String(), nil
}
